server/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -154,13 +161,53 @@ private static void collectTargetShardsWithRouting(
         }
     }
 
-    private static void collectTargetShardsNoRouting(RoutingTable routingTable, String[] concreteIndices, Set<IndexShardRoutingTable> set) {
+    private static void collectTargetShardsNoRouting(ProjectState projectState, String[] concreteIndices, Set<IndexShardRoutingTable> set) {
         for (String index : concreteIndices) {
-            final IndexRoutingTable indexRoutingTable = indexRoutingTable(routingTable, index);
-            for (int i = 0; i < indexRoutingTable.size(); i++) {
-                set.add(indexRoutingTable.shard(i));
+            Iterator<IndexShardRoutingTable> iterator = allShardsReadyForSearch(projectState, index);
+            while (iterator.hasNext()) {
+                set.add(iterator.next());
+            }
+        }
+    }
+
+    /**
+     * Returns an iterator of shards of the index that are ready to execute search requests.
+     * A shard may not be ready to execute these operations during processes like resharding.
+     */
+    private static Iterator<IndexShardRoutingTable> allShardsReadyForSearch(ProjectState projectState, String index) {
+        return allShardsExceptSplitTargetsInStateBefore(projectState, index, IndexReshardingState.Split.TargetShardState.SPLIT);
+    }
+
+    /**
+     * Returns an iterator of shards of the index that are ready to execute write requests.
+     * A shard may not be ready to execute these operations during processes like resharding.
+     */
+    private static Iterator<IndexShardRoutingTable> allShardsReadyForWrites(ProjectState projectState, String index) {
```

Comment:
I'm struggling a bit with naming. I worry that `Ready` could be confused with a shard that's in recovery. Such a shard might refuse operations but operations should be routed there anyway, because it's the owning shard of a request. Maybe `allWriteAddressableShards`?

Reply:
I like that!